src/EFCore.SqlServer/Update/Internal/SqlServerUpdateSqlGenerator.cs

```diff
@@ -133,12 +196,73 @@ public class SqlServerUpdateSqlGenerator : UpdateSqlGenerator, ISqlServerUpdateS
             return ResultSetMapping.LastInResultSet;
         }
 
-        return AppendBulkInsertWithServerValues(
+        // We default to using MERGE ... OUTPUT (without INTO), projecting back a synthetic _Position column to know the order back
+        // at the client and propagate database-generated values correctly. However, if any triggers are defined, OUTPUT without INTO
+        // doesn't work.
+        if (!AffectsTableWithTriggers(firstCommand))
```

Comment:
Commands are already grouped by table.

```suggestion
        if (!firstCommand.Entries[0].EntityType.Model.GetRelationalModel().FindTable(firstCommand.TableName, firstCommand.Schema).Triggers.Any())
```

Reply:
Thanks...